server/internal/thirdparty/tool-metrics/roundtripper.go

```diff
@@ -0,0 +1,233 @@
+				h.logger.ErrorContext(ctx, "failed to read response body", attr.SlogError(readErr))
+			}
+
+			if n > 0 {
+				respBodyBuffer.Write(firstChunk[:n])
+			}
+
+			// Reset the body for the proxy to handle streaming
+			resp.Body = io.NopCloser(io.MultiReader(
+				bytes.NewReader(firstChunk[:n]),
+				resp.Body,
+			))
+		} else {
+			// For non-streaming responses, read the entire body
+			respBodyBytes, err = io.ReadAll(resp.Body)
+			if err != nil {
+				h.logger.ErrorContext(ctx, "failed to read response body", attr.SlogError(err))
+			}
+
+			resp.Body = io.NopCloser(bytes.NewReader(respBodyBytes))
+		}
+	}
+
+	// Get request headers
+	requestHeaders := make(map[string]string)
+	for key, values := range req.Header {
+		for _, value := range values {
+			requestHeaders[key] = value
+		}
+	}
+
+	// Get response headers
+	responseHeaders := make(map[string]string)
+	if resp != nil {
+		for key, values := range resp.Header {
+			for _, value := range values {
+				responseHeaders[key] = value
+			}
+		}
+	}
+
+	var requestBodySkip *string = nil // where will this come from?
+
+	// Prepare response body data for logging
+	var responseBody *string
+	var responseBodySkip *string
+	var responseBodyBytes uint64
+
+	if isStreaming {
+		// For streaming responses, mark as truncated
+		responseBody = conv.Ptr(respBodyBuffer.String())
+		responseBodySkip = nil
+		responseBodyBytes = uint64(len(respBodyBuffer.Bytes()))
+	} else {
+		// For non-streaming responses, log the full body
+		responseBody = conv.Ptr(string(respBodyBytes))
+		responseBodySkip = nil
+		responseBodyBytes = uint64(len(respBodyBytes))
+	}
+
+	// Log this individual HTTP attempt to ClickHouse
+	logErr := h.tcm.Log(ctx, ToolHTTPRequest{
+		Ts:                time.Now(),
+		OrganizationID:    tool.OrganizationID,
+		ProjectID:         tool.ProjectID,
+		DeploymentID:      tool.DeploymentID,
+		ToolID:            tool.ID,
+		ToolURN:           tool.Urn,
+		ToolType:          ToolTypeHttp,
+		TraceID:           traceID,
+		SpanID:            spanID,
+		HTTPMethod:        req.Method,
+		HTTPRoute:         req.URL.Path,
+		StatusCode:        statusCode,
+		DurationMs:        durationMs,
+		UserAgent:         req.UserAgent(),
+		ClientIPv4:        clientIP,
+		RequestHeaders:    requestHeaders,
+		RequestBody:       conv.Ptr(string(reqBodyBytes)),
+		RequestBodySkip:   requestBodySkip,
+		RequestBodyBytes:  uint64(len(reqBodyBytes)),
+		ResponseHeaders:   responseHeaders,
+		ResponseBody:      responseBody,
+		ResponseBodySkip:  responseBodySkip,
+		ResponseBodyBytes: responseBodyBytes,
+	})
+
+	if logErr != nil {
+		h.logger.ErrorContext(ctx, "failed to log HTTP attempt to ClickHouse",
+			attr.SlogError(logErr),
+			attr.SlogToolID(tool.ID),
+			attr.SlogHTTPRequestMethod(req.Method),
+			attr.SlogURLOriginal(req.URL.String()),
+		)
+	}
+
+	if isStreaming {
+		// Log that we handled a streaming response
+		h.logger.DebugContext(ctx, "logged streaming HTTP response to ClickHouse",
```

Comment:
Doesn't seem like we are handling a streaming response right now. Or is that just not added